### PR TITLE
Cscript

### PR DESCRIPTION
**Pull request by hexagonal-sun**
Add support for `cscript`.  A way to run C code that looks like a script.

### CODE REVIEW

**Comment by hexagonal-sun** [comment]
Need to add support for `argc` and `argv` passing.

**Comment by cjwelborn** [comment]
Just so you know, I didn't get a chance to look at it yesterday because I was feeling kinda bad, but I am  actually excited to start running some code through this thing to see what it can do. I'll get back to you ASAP if I come across anything. I can tell from the example code that you posted that it's already going to be awesome.

**Comment by hexagonal-sun** [comment]
Hey @cjwelborn ,

Just thought you'd like to know that I've now pushed argument passing support for cscript into this branch.  For example, the following cscript:

```c
#include <stdio.h>

for (int i = 0; i < argc; i++)
        printf("argv[%d] = %s\n",i, argv[i]);
```
when executed with: `./src/bic -s test.csh foo  -t bar baz` produces:

```
argv[0] = test.csh
argv[1] = foo
argv[2] = -t
argv[3] = bar
argv[4] = baz
```

Hope this is useful.  Let me know if you're finding any issues. If not, I'll look at merging this into develop soon!

**Comment by cjwelborn** [comment]
@hexagonal-sun, I found an issue. I can't run your example because `stdio.h` can't be parsed:

deleteme.c:
```c
#include <stdio.h>

for (int i = 0; i < argc; i++)
        printf("argv[%d] = %s\n",i, argv[i]);
```

Running:
```bash
bic -s deleteme.c
```

Output:
```
Parser Error: /usr/include/stdio.h:135 syntax error, unexpected TYPE_NAME, expecting IDENTIFIER or '{'.
```
My `stdio.h` comes from `glibc`, and looks like [this](https://sourceware.org/git/?p=glibc.git;a=blob;f=libio/stdio.h;h=5a10a4d3ee94e8d960c6a2dadbea4355cccab77e;hb=768c83b7f60d82db6677e19dc51be9f341e0f3fc#l135) (there are tiny differences in the naming of `FILE`, I tried to find the exact file without uploading mine). It seems to choke right after that include ([bits/stdio_lim.h](https://sourceware.org/git/?p=glibc.git;a=blob;f=stdio-common/stdio_lim.h.in;h=de0caa5feef89c6f57221fa58e822869766cd5b4;hb=768c83b7f60d82db6677e19dc51be9f341e0f3fc)), but I don't see anything major there (this is the final version of `stdio_lim.h` for me):
```c
#ifndef _BITS_STDIO_LIM_H
#define _BITS_STDIO_LIM_H 1

#ifndef _STDIO_H
# error "Never include <bits/stdio_lim.h> directly; use <stdio.h> instead."
#endif

#define L_tmpnam 20
#define TMP_MAX 238328
#define FILENAME_MAX 4096

#ifdef __USE_POSIX
# define L_ctermid 9
# if !defined __USE_XOPEN2K || defined __USE_GNU
#  define L_cuserid 9
# endif
#endif

#undef  FOPEN_MAX
#define FOPEN_MAX 16

#endif /* bits/stdio_lim.h */
```

My line `135` looks like this (names are different):
```c
extern struct _IO_FILE *stdin;		/* Standard input stream.  */
```

Unless, those preprocessor directives are causing it, bic having it's own parser and all? I'll try to take a look at it later. I just had a minute, and I was excited to try this out. I'm running `Kubuntu 18.04. I know you already had to make a change because `libm` was handled differently on your machine. This may be another one of those compatibility issues. I'll let you know if I find anything out.

**Comment by cjwelborn** [comment]
I can't use `stdio`, but basic functionality is there :).

**Comment by hexagonal-sun** [comment]
Hey @cjwelborn ,

Thanks for the report.

I think I've already ran into this issue and I added a fix into #34 PR.  I'll rebase this branch on top of that to include the fix.

**Comment by cjwelborn** [comment]
Cool, I thought this was going to be another can of worms. I'm gonna try again when I get home.

**Comment by cjwelborn** [comment]
@hexagonal-sun, I've been messing around with this and it's awesome. There was one weird thing I found, that may come as a surprise to people using it. `return` doesn't work when nested in an `if` statement:

```c
#include <stdio.h>

if (argc < 2) {
    puts("No arguments!");
    return 1;
}

printf("Arguments: %d\n", argc);
return 0;
```

Running `bic -s myfile.c` results in both lines being printed:
```
No arguments!
Arguments: 1
```

You have to wrap the other case in an `else` to get the desired behavior.

Everything else I've tried so far works perfectly. Defining functions and calling them at the top level, using a function call for the final `return` (`return my_function();`). It's really cool. I have to figure out how to tell my linter that `argc` and `argv` are already defined. If I write in this style:
```c
#include <stdio.h>

int add2(int x) {
    return x + 2;
}

int main() {
    int y = add2(3);
    printf("%d\n", y);
    return 0;
}

return main();
```

...where only the last line is incompatible with traditional C, I can use my linter for almost everything. That's more of a tooling issue though. I'll probably be messing around with `bic` and this new feature a lot tomorrow.

**Comment by cjwelborn** [comment]
If this ends up opening the door to a whole bunch of compatibly problems or changes to `bic`, I was wondering if restricting everything to traditional C and just calling `main(argc, argv)` when one exists would be easier? Then without `main`, it drops into the REPL? I don't know enough about bic itself to make that call right now. I love that this top-level stuff works like a scripting language. I was just thinking about C versus "bic's C", and how different they may become in the future.

**Comment by hexagonal-sun** [comment]
Hey @cjwelborn ,

Thanks for reporting the bug with a return from `if`.  I'll look at fixing that... it has to do with the way I implement a return from a function call and since there is no actual function call happening, it's not returning!

> If this ends up opening the door to a whole bunch of compatibly problems or changes to bic, I was wondering if restricting everything to traditional C and just calling main(argc, argv) when one exists would be easier?

I don't think there should be too many other issues if I'm honest.  The changes that are required for this to work are quite minimal, and the grammar didn't turn out to be ambiguous which was my biggest worry.  I think if we fix the return from `if` there shouldn't be much else to fix.

> Then without main, it drops into the REPL?

Not a bad idea! I suppose we have two options if we drop support for top-level code:

- Drop to a REPL as you suggested
- Raise an error saying that there is no `main` function defined.

What do you think?

> I don't know enough about bic itself to make that call right now. I love that this top-level stuff works like a scripting language. I was just thinking about C versus "bic's C", and how different they may become in the future.

Indeed, I've been pondering whether `cscript` should be a thing, or whether we stick to traditional C now that I've implemented argc and argv parsing. Some pros and cons or merging cscript:

**Pros**
- Works like majority of today's scripting languages - code at the top-level doesn't have to be wrapped in a function.
- Reduces ambiguity if `main` isn't defined, the code is just executed.  If there aren't any calls to any functions, nothing happens.
- The entry point to the program doesn't have to be called `main`.

**Cons**
- Not standards compliment
- Tooling (editors, linters, etc..) that work with C won't work with cscript. 
- Requires yet another parser and lexer to be generated - this has to be maintained.
- cscript files can't be compiled.

That's just having a quick five minute ponder.  Maybe we carry on with the PR for now and see how it develops and whether we find many more issues with implementing `cscript` in this way?

**Comment by cjwelborn** [comment]
After reading your points I started thinking about how `BASH`, `Python`, `Perl`, and others do it. `bic` was a REPL first, and a "script executor" second, so it won't be exactly the same.  `BASH` doesn't drop into a REPL when executing a script, but it does have startup files. `Python` is a REPL by default, will execute script files from top to bottom (if there is no early `exit()` or `Exception`), but has to be told with `-i` to drop into the REPL after executing commands or scripts. It also has an optional startup file for REPL-mode. If `bic` is a REPL by default, will execute scripts with `-s` and exit gracefully when it encounters a `return`, it's already like the other tools/languages. If it drops into a REPL when there is no `return`, it becomes something different. *I know I was the one asking for that*, but now I think it should stay simple and predictable. In the future, with another option (`-i`?), you could be explicit about when you're executing a script, and when you're pre-loading code for a REPL environment. The dedicated startup file (`~/.config/startup.bic`?) is a whole other thing too. They would probably share code, but the dedicated startup file usually doesn't load when executing scripts, only for REPL-mode. It's just something to consider, when everything else is ironed out.

They also don't require a specific function to be defined as an entry point. You just make your own by wrapping code in functions, and call it after all of your declarations/definitions. They don't care what you call it, or where you call it, it's just top-to-bottom. `bic` is doing that now, and it's probably the best way (after the early-`return` thing is fixed). 

I guess my hangup was that C is so different as a compiled language with it's rules for `main` as the entry point (I know you can do weird stuff to circumvent this, or use `_start` with `-nostartfiles`), and the way `argc` and `argv` are traditionally passed in. Honestly though, `bic` is already outside of "traditional C". We might as well consider a top-level `return` as syntactic sugar for `exit()`, or imagine the whole file wrapped inside a big `bic_main()` function definition that `bic` executes instead of `_start`. **I like it as is, and I'd like to see where it goes.**

Sorry for the long posts. I could talk about this stuff all day. **I know I'm not telling you anything you don't already know.** I'm just writing down my thoughts on it. I'm reading up on `bison`/`flex` as I'm messing around with `bic`, something I've always wanted to do. This project gave me another reason to go ahead and dive in. I hate even proposing an idea without code to show for it, so hopefully I'll have more to offer in the future. As always, I'll let you know if I find anything weird/broken.

**Comment by hexagonal-sun** [comment]
Thanks for the feedback @cjwelborn, it's much appreicated and it's good to get other people's perspective on things.

I've just pushed a fix for the nested return so that should now work.

> Honestly though, bic is already outside of "traditional C"

Before the `cscript` component of `bic`, I've always tried to keep it separated into two modes:

1. Executing C files, expecting a `main` function to be defined. I've tried to keep this mode as standards-compliant as practicable (there are some exceptions since this is interpreted, of course).  The thinking here is that the closer I get this to the standard the more useful the REPL will be and more programs could be ran interpreted rather than compiled.

2. The REPL.  This is where I'm allowing myself to "have a bit of fun". Since there is no standard, I'm essentially free to implement this as I would like.

I do see your point that since `bic` is interpreted it's not "traditional C", however I'm keen for the former mode to be as close to the standard as we can make it.  Having said that, I do wonder whether we could merge the former mode and `cscript` into a single mode for bic?  I think it could get confusing for `bic` to have too many modes of operation.  I think you're advocating this: bic only has two modes, `cscript` and the REPL, which I agree with **if** we can make it also execute files that can also be compiled.

Having read your comments and considering the above, I propose the following:

1. Remove the first mode mentioned above and merge this with `cscript` in the following way:
  1.1. If a `cscript` file is found to have a `main` function, that is called in the traditional way, passing    `argc` and `argv`.  This allows us to still be able to execute 'compilable' C files.
  1.2. If `main` is defined and top-level code has also been provided, error.
  1.3. If no `main` function is found, we execute any top-level code that exists.
  1.4. If no `main` exists and no top-level code exists `bic` will parse and evaluate the function definitions but not do anything and exit with a return-code of `0`.
2. Enable `cscript` mode by passing `-s` to bic along with the file to be executed.
3. If `-s` isn't passed to `bic` we always drop to a REPL with `argc` and `argv` undefined and ignore any arguments passed.
4. If `~/.bic` exists and we're going into a REPL, execute any code there (in `cscript` mode) before dropping into the REPL.  I'm unsure whether the code in `~/.bic` should be able to have visibility of `argc` and `argv`, what do you think?
5. For `cscript` mode, ignore `~/.bic`.

I don't think this is too much work and it should be a nice balance between keeping `bic` simple (i.e. only two modes) and flexible.  If you think the above is a reasonable approach I'll look at getting this implemented this week/weekend.

**Comment by hexagonal-sun** [comment]
Okay, I've not pushed code that will implement the above apart from 4 & 5. Let me know if you have any issues!

**Comment by cjwelborn** [comment]
Awesome, I'll check it out when I get home.

**Comment by hexagonal-sun** [comment]
Okay, I've not implemented the last feature, which is reading `~/.bic`.  I have created mine as:

```c
#include <stdio.h>
#include <unistd.h>

puts("Ready and waiting!");

int a = 1024;
```
and I can do:
```
matthew@laforge GIT~(cscript) ~/Development/bic $ ./src/bic
Ready and waiting!
BIC> a;
1024
BIC> 
```

**Comment by cjwelborn** [comment]
It already looks great, but I'm gonna try some stuff tomorrow. It's gonna be cool to have a C repl with a few helper functions already defined.

**Comment by cjwelborn** [comment]
Getting a segfault with a `~/.bic` file like this (just testing some stuff out):
```c
/*
    Startup file for bic:
        github.com/hexagonal-sun/bic

    -Christopher Welborn 1-26-20
*/

int gt(int x, int y) {
    return x > y ? 1 : 0;
}
```

It's [here, in evaluate.c:340](https://github.com/hexagonal-sun/bic/blob/95d6a89f5e9b00e1a8a0d370a86bc610c11b5484/src/evaluate.c#L340). `evaluated` is `NULL` after this call on line [337](https://github.com/hexagonal-sun/bic/blob/95d6a89f5e9b00e1a8a0d370a86bc610c11b5484/src/evaluate.c#L337):
```c
tree evaluated = __evaluate_1(t, depth + 1),
        ret;
```

**Comment by hexagonal-sun** [comment]
@cjwelborn: the latest commits should fix your startup file.  I've refactored the top-level execution code.  I've made `__evaluate` return `void` since it's not clear what it should return when evaluating a function definition, say.

**Comment by cjwelborn** [comment]
Cool, I've been a little busy lately so I haven't got to sit down and mess with this stuff. I was hoping to fix that PR Saturday/Sunday, and play around with the startup file (and other script sruff) some more.

File: src/cscript.c [hunk 1]
```diff
@@ -179,7 +179,7 @@ static tree __evaluate_cscript(tree head, const char *fname, int argc,
     if (has_main)
         add_call_to_main(head, num_main_args);
 
-    return evaluate(head, fname);
+    evaluate(head, fname);
```
Comment (cjwelborn) [inline src/cscript.c]: Control may reach the end of a non-`void` function here. The signature is:
```c
static tree __evaluate_cscript()
```

..but it returns nothing.

**Comment by cjwelborn** [comment]
Okay, so I fixed that little PR I sent you, the startup file is working (awesome!), but I uncovered a segfault in the "inspector". This may need a separate issue, since it's not `cscript` related (i think). This is what I see (I ran it through `valgrind` to gain some insight):
```
BIC> int x = 5;
x
BIC> ?x
x is a signed integer.
value of x is 5
sizeof(x) = 4 bytes.
==31644== Invalid read of size 4
==31644==    at 0x123CA6: tree_check (tree.h:105)
==31644==    by 0x123CA6: print_object_locus (inspect.c:65)
==31644==    by 0x123CA6: inspect (inspect.c:74)
==31644==    by 0x116147: eval_inspect (evaluate.c:1030)
==31644==    by 0x116147: __evaluate_1 (evaluate.c:2699)
==31644==    by 0x12155F: bic_repl (repl.c:469)
==31644==    by 0x10FAA5: main (main.c:162)
==31644==  Address 0x0 is not stack'd, malloc'd or (recently) free'd
```

Another null pointer, but I haven't tracked it down yet. I'm looking. It looks like the `tree` may be `NULL` when using the `TYPE(obj)` macro call (similar to before). I figured you might recognize what went wrong before I do.

**Comment by hexagonal-sun** [comment]
@cjwelborn, your issue should now be fixed.  Looks like I wasn't setting the locus of `IDENTIFIERS` in the new parser!

**Pull request merged**